### PR TITLE
Validate all relevant square sizes are committed to in `MsgWirePayForData`

### PR DESCRIPTION
**Pull request by rootulp**
Closes https://github.com/celestiaorg/celestia-app/issues/620 (after https://github.com/celestiaorg/celestia-app/pull/654 merges)

### CODE REVIEW

File: x/payment/types/payfordata_test.go
Comment (rootulp) [inline x/payment/types/payfordata_test.go:422]: [for reviewers]
Updated this test data generator because a message with 2000 bytes expects all these square sizes. This intuitively makes sense b/c IIUC the algorithm that defines which square sizes must be signed over starts at max (128) and proceeds by lower powers of two until a square size is lower than the message size.

In this case 2000 bytes can't fit in a square size of 2 x 2 x 256 = 1,024 bytes so 4 is the lowest square needed.

File: x/payment/types/wirepayfordata.go
Comment (rootulp) [inline x/payment/types/wirepayfordata.go:148]: [for reviewers] we don't need to use `reflect.DeepEqual` here but I used it for simplicity. If performance is critical on this code path then we can consider using a custom Equals function:

Ref: https://yourbasic.org/golang/compare-slices/

File: x/payment/types/wirepayfordata.go
Comment (rootulp) [inline x/payment/types/wirepayfordata.go:146]: [for reviewers] sorting isn't strictly necessary here. We could convert these slices to sets (perhaps backed by a `map[uint64]bool`) and implement an equality operator for these sets. 

Performance would improve from O(N log N) for sorting to O(N) b/c constant time operations for all elements in `allSquareSizes` + `committedSquareSizes`. This type of optimization seems premature given the length of `allSquareSizes` is currently capped to 8. `commitedSquareSizes` is potentially unbounded so we should consider adding a check for that being too large 😳

**Comment by rootulp** [comment]
I expect merge conflicts with https://github.com/celestiaorg/celestia-app/pull/654 so I'd like to merge that PR in first, prior to this one.

File: x/payment/types/wirepayfordata_test.go
Comment (rootulp) [inline x/payment/types/wirepayfordata_test.go:50]: [for reviewers] this test had to be updated to use a square size that is not 4 because after this line change:
https://github.com/celestiaorg/celestia-app/pull/666/files#diff-0535627976b096bde128716f4bbe48b835785e957a6466740da9fb8e6fe1f967R422

the first square size is `4` so `badSquareSizeMsg.MessageShareCommitment[0].K = 4` was a no-op. It took me a silly long time to realize why I broke an unrelated test while working on this

File: x/payment/types/wirepayfordata.go
Comment (evan-forbes) [reply x/payment/types/wirepayfordata.go:148]: given the simplicity of using something else, imo I think we should. While the performance gained or lost is arguably trivial compared to checking that all of the commitments are actually accurate, I still think we should scavenge as much performance as possible given https://github.com/informalsystems/audit-celestia/issues/6

Also, even a small percentage will add up over time since this function will be called by every full node for every message submitted.

File: x/payment/types/wirepayfordata.go
Comment (rootulp) [reply x/payment/types/wirepayfordata.go:149]: > is potentially unbounded so we should consider adding a check for that being too large 

Resolved in https://github.com/celestiaorg/celestia-app/pull/666/commits/baa170ba27f64164ec2b9cc5ef446723e3547312

**Comment by rootulp** [comment]
Sorry for prematurely moving this from draft to ready for review. Tests passed locally but CI is failing on `TestIntegrationTestSuite` which presumably doesn't run locally. Will fix and tag reviewers when this PR is ready.

**Review by evan-forbes — COMMENTED** [review]
left a few comments :slightly_smiling_face: 

this code is really important because it makes arranging the square way easier when the block proposer knows that all the txs in their mempool are signed over all the relevant square sizes.

File: x/payment/types/wirepayfordata.go
Comment (rootulp) [reply x/payment/types/wirepayfordata.go:153]: Resolving in favor of https://github.com/celestiaorg/celestia-app/pull/666#discussion_r959495851

File: app/test/prepare_proposal_test.go
Comment (rootulp) [reply app/test/prepare_proposal_test.go:39]: `AllSquareSizes` is defined [here](https://github.com/rootulp/celestia-app/blob/13076a62c567f624b2508752b7e129467cd71bf9/x/payment/types/square_sizes.go#L26). It is possible to use it here. I opted to retain existing behavior to avoid bloating this PR with a refactor. Happy to update in this PR (if using `AllSquareSizes` is blocking) or in a in a follow-up PR (if non-blocking).

**Note**: I don't see a corresponding `square_sizes_test.go` but will create an issue for it. I think it's important to unit test the behavior of `AllSquareSizes` if we're going to update all tests to using it.

File: x/payment/types/wirepayfordata.go
Comment (rootulp) [reply x/payment/types/wirepayfordata.go:181]: > why not pre-define the slice capacity here?

By default I favor readability over performance. When one needs to optimize for performance, they can evaluate with benchmarks and selectively make the existing readable code more performant where needed. I think the code above is _slightly_ more readable than the following but I understand this is subjective.

```golang
func (msg *MsgWirePayForData) committedSquareSizes() []uint64 {
	squareSizes := make([]uint64, 0, len(msg.MessageShareCommitment))
	...
}
```

> I think tendermint goes to the extent to include this in the linter.

You're right. Looks like:
- https://github.com/tendermint/tendermint/blob/f1835338d9f44d7bc215d84ff0c72a49c61897d3/.golangci.yml#L22
- https://golangci-lint.run/usage/linters/#prealloc

I'm in favor of enabling a linter like this to: (a) reduce PR feedback cycles and (b) build more codebase consistency. 

**Note**: the linter config says

```
    # IMPORTANT: we don't recommend using this linter before doing performance profiling.
    # For most programs usage of prealloc will be a premature optimization.
```

I attempted to enable locally but it didn't flag the instance in this PR 

```shell
$ golangci-lint run --verbose
...
INFO [lintersdb] Active 15 linters: [deadcode errcheck exportloopref gofumpt gosimple govet ineffassign misspell prealloc revive staticcheck structcheck typecheck unused varcheck]
...
// no issues reported
```

To avoid blocking this PR on this nit, will pre-allocate and create an issue to enable the linter (after confirming it works)

Comment (rootulp) [reply x/payment/types/wirepayfordata.go:182]: Resolved in https://github.com/celestiaorg/celestia-app/pull/666/commits/30b8e8e66734a4f6a75e980f4a4ec895ec09328e

Comment (rootulp) [reply x/payment/types/wirepayfordata.go:182]: Created https://github.com/celestiaorg/celestia-app/issues/676

File: x/payment/types/wirepayfordata.go
Comment (rootulp) [reply x/payment/types/wirepayfordata.go:154]: Created https://github.com/celestiaorg/celestia-specs/issues/213

Comment (rootulp) [reply x/payment/types/wirepayfordata.go]: Removed sort in https://github.com/celestiaorg/celestia-app/pull/666/commits/639be1c84c7eb1a395e4822a43013f8f0c74165b

Comment (evan-forbes) [reply x/payment/types/wirepayfordata.go]: > I was not aware of a validity rule for share commitments being presented

I mean, there wasn't one!! :sweat_smile: 

what you did was correct :slightly_smiling_face:

File: app/test/prepare_proposal_test.go
Comment (evan-forbes) [reply app/test/prepare_proposal_test.go:39]: > I opted to retain existing behavior to avoid bloating this PR with a refactor

> I don't see a corresponding square_sizes_test.go but will create an issue for it. I think it's important to unit test the behavior of AllSquareSizes if we're going to update all tests to using it.

I understand not wanting to use it if there are no unit tests, so I'm not going to block on this PR. I do however think it falls under the scope of #620. Since we are adding a new validity rule that relies on a function to calculate square sizes. It seems reasonable to also use that same function to generate tests.

if not, pls create an issue with the current context before merging. Thanks!

Comment (rootulp) [reply app/test/prepare_proposal_test.go]: > It seems reasonable to also use that same function to generate tests.

Agreed. Resolved in https://github.com/celestiaorg/celestia-app/pull/666/commits/e5f55a043174b324b20bc2df04f6fdfad72490c0

**Comment by evan-forbes** [comment]
just to note that this is consensus breaking

**Review by evan-forbes — APPROVED** [review]
approving due to #677

**Comment by evan-forbes** [comment]
I'm dumb, this is probably not consensus breaking

**Comment by rootulp** [comment]
I think this is consensus breaking b/c prior to this change, validators would've accepted PFDs that didn't sign over all commits. After this change, those PFDs are considered invalid. 

Consensus tests would be very nice. +1 to https://github.com/celestiaorg/celestia-app/issues/313

**Comment by rootulp** [comment]
Correct only the wire message, sorry for the lack of precision.

**Comment by evan-forbes** [comment]
Then that _should_ only affect the mempool, as even if a `WireMsg` makes it in the block, then all nodes should treat that the same, which is to subtract the minimum amount of gas.

**Pull request merged**
